[autograd.Function] add nice error message for incorrect usage of vmap

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #92030
* #92026
* #92025
* #92024
* __->__ #92023
* #91787

This PR:
- adds a nice error message if the user doesn't follow the API of the
  vmap staticmethod correctly. That is, the user must return two
  arguments from the vmap staticmethod API: (outputs, out_dims), and
  out_dims must be a PyTree with either the same structure as `outputs`
  our be broadcastable to the same structure as `outputs`.
- Fixes an edge case for out_dims=None. out_dims is allowed to be None,
  but wrap_outputs_maintaining_identity was treating "None" as "This is
  not the vmap case"

Test Plan:
- new tests

cc @gujinghui @PenghuiCheng @XiaobingSuper @jianyuh @jgong5 @mingfeima @sanchitintel @ashokei @jingxu10 @min-jean-cho @yanbing-j @Guobing-Chen @Xia-Weiwen

zou3519 committed:
[autograd.Function] add nice error message for incorrect usage of vmap

This PR:
- adds a nice error message if the user doesn't follow the API of the
  vmap staticmethod correctly. That is, the user must return two
  arguments from the vmap staticmethod API: (outputs, out_dims), and
  out_dims must be a PyTree with either the same structure as `outputs`
  our be broadcastable to the same structure as `outputs`.
- Fixes an edge case for out_dims=None. out_dims is allowed to be None,
  but wrap_outputs_maintaining_identity was treating "None" as "This is
  not the vmap case"

Test Plan:
- new tests

[ghstack-poisoned]

zou3519 committed:
Update on "[autograd.Function] add nice error message for incorrect usage of vmap"


This PR:
- adds a nice error message if the user doesn't follow the API of the
  vmap staticmethod correctly. That is, the user must return two
  arguments from the vmap staticmethod API: (outputs, out_dims), and
  out_dims must be a PyTree with either the same structure as `outputs`
  our be broadcastable to the same structure as `outputs`.
- Fixes an edge case for out_dims=None. out_dims is allowed to be None,
  but wrap_outputs_maintaining_identity was treating "None" as "This is
  not the vmap case"

Test Plan:
- new tests

[ghstack-poisoned]

zou3519 committed:
Update on "[autograd.Function] add nice error message for incorrect usage of vmap"


This PR:
- adds a nice error message if the user doesn't follow the API of the
  vmap staticmethod correctly. That is, the user must return two
  arguments from the vmap staticmethod API: (outputs, out_dims), and
  out_dims must be a PyTree with either the same structure as `outputs`
  our be broadcastable to the same structure as `outputs`.
- Fixes an edge case for out_dims=None. out_dims is allowed to be None,
  but wrap_outputs_maintaining_identity was treating "None" as "This is
  not the vmap case"

Test Plan:
- new tests

[ghstack-poisoned]

soulitzer commented on torch/_functorch/autograd_function.py:
```diff
@@ -267,16 +291,14 @@ def custom_function_call_vmap(interpreter, autograd_function, *operands):
     with interpreter.lower():
-        unwrapped_output, out_dims = autograd_function.vmap(info, in_dims, *unwrapped_operands)
+        result = autograd_function.vmap(info, in_dims, *unwrapped_operands)
+    validate_vmap_returns_two(result)
```
nit: I think we might as well fully describe it here, then a reader won't even have to read the function to figure out what it does - "validate_vmap_returns_tuple_of_two_elements".

soulitzer commented on test/functorch/test_eager_transforms.py:
```diff
@@ -1295,6 +1295,136 @@ def f(x):
+            @staticmethod
+            def vmap(info, in_dims, input):
+                assert in_dims == (0,)
+                return torch.zeros(input.shape[1:], device=input.device), None
```
Why does this work? Does _broadcast_to_and_flatten do something special if out_dims is None? Otherwise it seems that the wrap_fn we define should just not wrap the output in a batched tensor.

zou3519 replied:
Your reading is correct -- `_broadcast_to_and_flatten ` doesn't have any special logic for None. It is intended that vmap does not construct a BatchedTensor in this case. 

The expected behavior of this is that the output (call it `zeros`) is not wrapped in a BatchedTensor. That's OK. Inside of `vmap`, a Tensor is either a BatchedTensor or a regular Tensor. This is the latter case.

There are two cases:
- Eventually the regular Tensor interacts with a BatchedTensor. E.g. `regular_tensor * batched_tensor`. This produces a BatchedTensor.
- The regular Tensor gets returned from the function being vmapped over. vmap has special handling to take this Tensor and expand out the batch dimension.

soulitzer commented on torch/_functorch/autograd_function.py:
```diff
@@ -145,11 +145,14 @@ def jvp(ctx, *tangents):
     )
     return Generated
 
+NO_OUT_DIMS = "not specified"
```
This is to avoid confusing None with "not specified" right? Maybe we could have a comment about what happens in the None case?

soulitzer reviewed (approved):
Looks ok, just have a question about what happens in None case.

zou3519 committed:
Update on "[autograd.Function] add nice error message for incorrect usage of vmap"


This PR:
- adds a nice error message if the user doesn't follow the API of the
  vmap staticmethod correctly. That is, the user must return two
  arguments from the vmap staticmethod API: (outputs, out_dims), and
  out_dims must be a PyTree with either the same structure as `outputs`
  our be broadcastable to the same structure as `outputs`.
- Fixes an edge case for out_dims=None. out_dims is allowed to be None,
  but wrap_outputs_maintaining_identity was treating "None" as "This is
  not the vmap case"

Test Plan:
- new tests

[ghstack-poisoned]

yanbing-j commented:
Hi @zou3519 , do you need to make some changes in `third_party/ideep`? If not, perhaps you can sync the submodule to get the latest third_party.

zou3519 committed:
Update on "[autograd.Function] add nice error message for incorrect usage of vmap"


This PR:
- adds a nice error message if the user doesn't follow the API of the
  vmap staticmethod correctly. That is, the user must return two
  arguments from the vmap staticmethod API: (outputs, out_dims), and
  out_dims must be a PyTree with either the same structure as `outputs`
  our be broadcastable to the same structure as `outputs`.
- Fixes an edge case for out_dims=None. out_dims is allowed to be None,
  but wrap_outputs_maintaining_identity was treating "None" as "This is
  not the vmap case"

Test Plan:
- new tests

cc gujinghui @PenghuiCheng @XiaobingSuper jianyuh jgong5 mingfeima sanchitintel ashokei jingxu10 min-jean-cho yanbing-j @Guobing-Chen @Xia-Weiwen

[ghstack-poisoned]

zou3519 committed:
Update on "[autograd.Function] add nice error message for incorrect usage of vmap"


This PR:
- adds a nice error message if the user doesn't follow the API of the
  vmap staticmethod correctly. That is, the user must return two
  arguments from the vmap staticmethod API: (outputs, out_dims), and
  out_dims must be a PyTree with either the same structure as `outputs`
  our be broadcastable to the same structure as `outputs`.
- Fixes an edge case for out_dims=None. out_dims is allowed to be None,
  but wrap_outputs_maintaining_identity was treating "None" as "This is
  not the vmap case"

Test Plan:
- new tests

cc gujinghui @PenghuiCheng @XiaobingSuper jianyuh jgong5 mingfeima sanchitintel ashokei jingxu10 min-jean-cho yanbing-j @Guobing-Chen @Xia-Weiwen

[ghstack-poisoned]
